File: backend/app/routers/cycle.py
from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import JSONResponse
from sqlalchemy.orm import Session
from pydantic import BaseModel
from typing import Optional, List
from app.database import get_db
from app.models.cycle import CycleType, CycleVersion, CycleStep
from app.models.factory import Workstation, StorageLocation
from app.auth import require_admin, get_current_user
from app.services.cycle_service import create_new_version, export_cycle

router = APIRouter(prefix="/api/cycles", tags=["cycles"])


def step_out(s: CycleStep) -> dict:
    return {
        "id": s.id,
        "step_number": s.step_number,
        "step_order": s.step_order,
        "operation_name": s.operation_name,
        "workstation_id": s.workstation_id,
        "workstation_code": s.workstation.code if s.workstation else None,
        "workstation_name": s.workstation.name if s.workstation else None,
        "from_storage_id": s.from_storage_id,
        "from_storage_code": s.from_storage.code if s.from_storage else None,
        "to_storage_id": s.to_storage_id,
        "to_storage_code": s.to_storage.code if s.to_storage else None,
        "is_converting_step": s.is_converting_step,
        "is_child_marking_step": s.is_child_marking_step,
        "is_qc_step": s.is_qc_step,
    }


def version_out(v: CycleVersion) -> dict:
    return {
        "id": v.id,
        "version_number": v.version_number,
        "is_current": v.is_current,
        "created_at": v.created_at,
        "change_notes": v.change_notes,
        "steps": [step_out(s) for s in v.steps],
    }


def cycle_out(c: CycleType) -> dict:
    current = next((v for v in c.versions if v.is_current), None)
    return {
        "id": c.id,
        "name": c.name,
        "letter_prefix": c.letter_prefix,
        "description": c.description,
        "is_active": c.is_active,
        "is_archived": c.is_archived,
        "current_version": version_out(current) if current else None,
        "version_count": len(c.versions),
    }


@router.get("/")
def list_cycles(db: Session = Depends(get_db), _=Depends(get_current_user)):
    return [cycle_out(c) for c in db.query(CycleType).filter(CycleType.is_archived == False).all()]


@router.get("/{cycle_id}")
def get_cycle(cycle_id: int, db: Session = Depends(get_db), _=Depends(get_current_user)):
    c = db.query(CycleType).filter(CycleType.id == cycle_id).first()
    if not c:
        raise HTTPException(status_code=404, detail="Cycle not found")
    return cycle_out(c)


@router.get("/{cycle_id}/versions")
def list_versions(cycle_id: int, db: Session = Depends(get_db), _=Depends(get_current_user)):
    versions = db.query(CycleVersion).filter(CycleVersion.cycle_type_id == cycle_id).order_by(CycleVersion.version_number.desc()).all()
    return [version_out(v) for v in versions]


class CycleCreate(BaseModel):
    name: str
    letter_prefix: str
    description: Optional[str] = None


@router.post("/", status_code=201)
def create_cycle(body: CycleCreate, db: Session = Depends(get_db), _=Depends(require_admin)):
    if db.query(CycleType).filter(CycleType.letter_prefix == body.letter_prefix).first():
        raise HTTPException(status_code=400, detail="Letter prefix already in use")
    c = CycleType(name=body.name, letter_prefix=body.letter_prefix.upper(), description=body.description)
    db.add(c)
    db.commit()
    db.refresh(c)
    return cycle_out(c)


class StepData(BaseModel):
    step_number: str
    operation_name: str
    workstation_id: int
    from_storage_id: Optional[int] = None
    to_storage_id: Optional[int] = None
    is_converting_step: bool = False
    is_child_marking_step: bool = False
    is_qc_step: bool = False


class VersionCreate(BaseModel):
    steps: List[StepData]
    change_notes: Optional[str] = None


@router.post("/{cycle_id}/versions", status_code=201)
def create_version(
    cycle_id: int,
    body: VersionCreate,
    db: Session = Depends(get_db),
    user=Depends(require_admin),
):
    version = create_new_version(
        db,
        cycle_type_id=cycle_id,
        steps_data=[s.model_dump() for s in body.steps],
        created_by_id=user.id,
        change_notes=body.change_notes,
    )
    return version_out(version)


@router.get("/{cycle_id}/export")
def export(cycle_id: int, version_id: Optional[int] = None, db: Session = Depends(get_db), _=Depends(get_current_user)):
    data = export_cycle(db, cycle_id, version_id)
    return JSONResponse(content=data, headers={"Content-Disposition": f'attachment; filename="cycle_{data["cycle_name"]}_v{data["version_number"]}.json"'})


class CycleImport(BaseModel):
    data: dict   # raw exported JSON
    update_existing: bool = False


@router.post("/import", status_code=201)
def import_cycle(body: CycleImport, db: Session = Depends(get_db), user=Depends(require_admin)):
    data = body.data
    cycle_name = data.get("cycle_name")
    letter_prefix = data.get("cycle_letter_prefix")
    steps_raw = data.get("steps", [])

    existing = db.query(CycleType).filter(CycleType.name == cycle_name).first()
    if existing and not body.update_existing:
        raise HTTPException(status_code=400, detail=f"Cycle '{cycle_name}' already exists. Set update_existing=true to create a new version.")

    # Resolve workstation and storage codes to IDs
    steps_data = []
    for s in steps_raw:
        ws = db.query(Workstation).filter(Workstation.code == s["workstation_code"]).first()
        from_storage = db.query(StorageLocation).filter(StorageLocation.code == s.get("from_storage_code")).first() if s.get("from_storage_code") else None
        to_storage = db.query(StorageLocation).filter(StorageLocation.code == s.get("to_storage_code")).first() if s.get("to_storage_code") else None
        steps_data.append({
            **s,
            "workstation_id": ws.id if ws else None,
            "from_storage_id": from_storage.id if from_storage else None,
            "to_storage_id": to_storage.id if to_storage else None,
        })

    if not existing:
        existing = CycleType(name=cycle_name, letter_prefix=letter_prefix, description=f"Imported from file")
        db.add(existing)
        db.flush()

    version = create_new_version(db, existing.id, steps_data, user.id, f"Imported from file")
    return {"cycle_id": existing.id, "version_id": version.id, "version_number": version.version_number}

File: backend/src/routes/cycle.js
import { Router } from 'express';
import { query, one } from '../db/pool.js';
import { asyncHandler } from '../middleware/error.js';
import { requireAuth, requireAdmin, HttpError } from '../middleware/auth.js';
import { createNewVersion, exportCycle } from '../services/cycleService.js';

const router = Router();

async function stepsForVersion(versionId) {
  const rows = await query(
    `SELECT cs.id, cs.step_number, cs.step_order, cs.operation_name,
            cs.workstation_id, w.code AS workstation_code, w.name AS workstation_name,
            cs.from_storage_id, fs.code AS from_storage_code,
            cs.to_storage_id, ts.code AS to_storage_code,
            cs.is_converting_step, cs.is_child_marking_step, cs.is_qc_step,
            cs.capacity_per_unit,
            (SELECT COUNT(*)::int FROM workstation_units wu
              WHERE wu.workstation_id = cs.workstation_id AND wu.status = 'active') AS active_units
       FROM cycle_steps cs
       LEFT JOIN workstations w ON w.id = cs.workstation_id
       LEFT JOIN storage_locations fs ON fs.id = cs.from_storage_id
       LEFT JOIN storage_locations ts ON ts.id = cs.to_storage_id
      WHERE cs.cycle_version_id = $1
      ORDER BY cs.step_order`,
    [versionId]
  );
  return rows.map((r) => ({
    ...r,
    total_capacity: r.capacity_per_unit != null ? r.capacity_per_unit * r.active_units : null,
  }));
}

async function versionOut(v) {
  return {
    id: v.id,
    version_number: v.version_number,
    is_current: v.is_current,
    created_at: v.created_at,
    change_notes: v.change_notes,
    steps: await stepsForVersion(v.id),
  };
}

async function cycleOut(c) {
  const current = await one(
    'SELECT * FROM cycle_versions WHERE cycle_type_id = $1 AND is_current = TRUE',
    [c.id]
  );
  const countRow = await one('SELECT COUNT(*)::int AS n FROM cycle_versions WHERE cycle_type_id = $1', [c.id]);
  return {
    id: c.id,
    name: c.name,
    letter_prefix: c.letter_prefix,
    description: c.description,
    is_active: c.is_active,
    is_archived: c.is_archived,
    current_version: current ? await versionOut(current) : null,
    version_count: countRow.n,
  };
}

router.get(
  '/',
  requireAuth,
  asyncHandler(async (req, res) => {
    const rows = await query('SELECT * FROM cycle_types WHERE is_archived = FALSE ORDER BY id');
    res.json(await Promise.all(rows.map(cycleOut)));
  })
);

router.post(
  '/',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const { name, letter_prefix, description = null } = req.body || {};
    const dup = await one('SELECT id FROM cycle_types WHERE letter_prefix = $1', [letter_prefix.toUpperCase()]);
    if (dup) throw new HttpError(400, 'Letter prefix already in use');
    const c = await one(
      'INSERT INTO cycle_types (name, letter_prefix, description) VALUES ($1,$2,$3) RETURNING *',
      [name, letter_prefix.toUpperCase(), description]
    );
    res.status(201).json(await cycleOut(c));
  })
);

router.post(
  '/import',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const { data, update_existing = false } = req.body || {};
    const cycleName = data.cycle_name;
    const letterPrefix = data.cycle_letter_prefix;
    const stepsRaw = data.steps || [];

    let existing = await one('SELECT * FROM cycle_types WHERE name = $1', [cycleName]);
    if (existing && !update_existing) {
      throw new HttpError(400, `Cycle '${cycleName}' already exists. Set update_existing=true to create a new version.`);
    }

    const stepsData = [];
    for (const s of stepsRaw) {
      const ws = await one('SELECT id FROM workstations WHERE code = $1', [s.workstation_code]);
      const fromS = s.from_storage_code ? await one('SELECT id FROM storage_locations WHERE code = $1', [s.from_storage_code]) : null;
      const toS = s.to_storage_code ? await one('SELECT id FROM storage_locations WHERE code = $1', [s.to_storage_code]) : null;
      stepsData.push({
        ...s,
        workstation_id: ws ? ws.id : null,
        from_storage_id: fromS ? fromS.id : null,
        to_storage_id: toS ? toS.id : null,
      });
    }

    if (!existing) {
      existing = await one(
        'INSERT INTO cycle_types (name, letter_prefix, description) VALUES ($1,$2,$3) RETURNING *',
        [cycleName, letterPrefix, 'Imported from file']
      );
    }
    const versionId = await createNewVersion(existing.id, stepsData, req.user.id, 'Imported from file');
    const version = await one('SELECT * FROM cycle_versions WHERE id = $1', [versionId]);
    res.status(201).json({ cycle_id: existing.id, version_id: version.id, version_number: version.version_number });
  })
);

router.get(
  '/:cycleId',
  requireAuth,
  asyncHandler(async (req, res) => {
    const c = await one('SELECT * FROM cycle_types WHERE id = $1', [parseInt(req.params.cycleId, 10)]);
    if (!c) throw new HttpError(404, 'Cycle not found');
    res.json(await cycleOut(c));
  })
);

router.get(
  '/:cycleId/versions',
  requireAuth,
  asyncHandler(async (req, res) => {
    const rows = await query(
      'SELECT * FROM cycle_versions WHERE cycle_type_id = $1 ORDER BY version_number DESC',
      [parseInt(req.params.cycleId, 10)]
    );
    res.json(await Promise.all(rows.map(versionOut)));
  })
);

router.post(
  '/:cycleId/versions',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const cycleId = parseInt(req.params.cycleId, 10);
    const { steps = [], change_notes = null } = req.body || {};
    const versionId = await createNewVersion(cycleId, steps, req.user.id, change_notes);
    const version = await one('SELECT * FROM cycle_versions WHERE id = $1', [versionId]);
    res.status(201).json(await versionOut(version));
  })
);

router.get(
  '/:cycleId/export',
  requireAuth,
  asyncHandler(async (req, res) => {
    const versionId = req.query.version_id ? parseInt(req.query.version_id, 10) : null;
    const data = await exportCycle(parseInt(req.params.cycleId, 10), versionId);
    res.set('Content-Disposition', `attachment; filename="cycle_${data.cycle_name}_v${data.version_number}.json"`);
    res.json(data);
  })
);

export default router;
